sfxclient/multitokensink.go

```diff
@@ -0,0 +1,515 @@
+)
+
+type msg struct {
+	token string
+	data  interface{}
+}
+
+// stopPMsg is a "poison pill" used to stop the pipeline
+var stopMsg = &msg{
+	token: "STOPITNOW",
+}
+
+type atomicBool struct {
+	val  bool
+	lock *sync.Mutex
+}
+
+func (a *atomicBool) Get() bool {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	return a.val
+}
+
+func (a *atomicBool) Set(in bool) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	a.val = in
+}
+
+func newAtomicBool() *atomicBool {
+	return &atomicBool{
+		lock: &sync.Mutex{},
+	}
+}
+
+// worker manages a pipeline for emitting metrics
+type worker struct {
+	lock         *sync.Mutex       // lock to control concurrent access to the worker
+	input        chan *msg         // channel for inputing datapoints into a worker
+	maxBuffer    int               // maxBuffer is the maximum size of the buffer for data
+	errorHandler func(error) error // error handler for handling error emitting datapoints
+	dataBuffered int64             // atomic Int64 that maintains count of data in the sink
+	sink         *HTTPSink         // sink is an HTTPSink for emitting datapoints to Signal Fx
+	done         chan bool         // done is a boolean channel for signaling that the worker has stopped
+}
+
+// emits a series of datapoints
+func (w *worker) emit(token string, datapoints []*datapoint.Datapoint, events []*event.Event) {
+	// set the token on the HTTPSink
+	w.sink.AuthToken = token
+
+	if len(datapoints) > 0 {
+		// emit datapoints and handle any errors
+		if err := w.sink.AddDatapoints(context.Background(), datapoints); err != nil {
+			_ = w.errorHandler(err)
+		}
+		// account for the emitted datapoints
+		_ = atomic.AddInt64(&w.dataBuffered, int64(len(datapoints)*-1))
+	}
+	if len(events) > 0 {
+		// emit datapoints and handle any errors
+		if err := w.sink.AddEvents(context.Background(), events); err != nil {
+			_ = w.errorHandler(err)
+		}
+
+		// account for the emitted events
+		_ = atomic.AddInt64(&w.dataBuffered, int64(len(events)*-1))
+	}
+}
+
+// Stop will stop the pipeline
+func (w *worker) Stop(timeout time.Duration) bool {
+	for {
+		select {
+		case <-time.After(timeout): // timeout and generate error message indicating what may be dropped
+			return false
+		case w.input <- stopMsg:
+			// stop datapoint emission
+		case <-w.done:
+			defer close(w.done)
+			return true
+		}
+	}
+}
+
+// returns a new instance of worker with an configured emission pipeline
+func newWorker(maxBuffer int, errorHandler func(error) error) *worker {
+	w := &worker{
+		lock:         &sync.Mutex{},
+		sink:         NewHTTPSink(),
+		maxBuffer:    maxBuffer,
+		errorHandler: errorHandler,
+		done:         make(chan bool),
+	}
+
+	return w
+}
+
+// worker for handling datapoints
+type datapointWorker struct {
+	*worker
+	buffer []*datapoint.Datapoint
+}
+
+// processDatapointMsg buffers datapoints attached to datapoint pipeline messages and emits them when necessary
+func (w *datapointWorker) processMsg(msg *msg) (stop bool) {
```

Comment:
i really envisioned this whole thing being a lot simpler.  the out loop looks fine  with the exception of the extra variable and complexity.  

this method is basically here to chunk a large message into smaller if necessary. so i would expect first you would simply check if the data fits into the buffer and if so, put it in. if it doesn't, put as much as possible into it, and send and repeat.   how about something like this:

func (w *datapointWorker) processMsg(msg *msg) (stop bool) {
	if msg == stopMsg {
		stop = true
	} else {
		var data = msg.data.([]*datapoint.Datapoint)
		for len(data) > 0 {
			msgLength := len(data)
			remainingBuffer := (w.maxBuffer - len(w.buffer))
			if msgLength > remainingBuffer {
				msgLength = remainingBuffer
			}
			w.buffer = append(w.buffer, data[:msgLength]...)
			data = data[:msgLength]
			if len(w.buffer) >= w.maxBuffer {
				w.emit(msg.token, w.buffer, nil)
				w.buffer = w.buffer[:0]
			}
		}
	}
	return
}